src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2298,4 +2298,496 @@ - (void)testSubscriptionPool
+    MTRDeviceController * controller = [self startControllerWithRootKeys:rootKeys
+                                                         operationalKeys:operationalKeys
+                                                                fabricID:fabricID
+                                                                  nodeID:nodeID
+                                                                 storage:storageDelegate
+                                                                   error:&error
+                                                       certificateIssuer:&certificateIssuer];
+    XCTAssertNil(error);
+    XCTAssertNotNil(controller);
+    XCTAssertTrue([controller isRunning]);
+
+    XCTAssertEqualObjects(controller.controllerNodeID, nodeID);
+
+    certificateIssuer.nextNodeID = deviceID;
+    [self commissionWithController:controller newNodeID:deviceID];
+
+    MTRDevice * device = [MTRDevice deviceWithNodeID:deviceID controller:controller];
+    return device;
+}
+
+- (NSMutableArray<NSNumber *> *)getEndpointArrayFromPartsList:(NSDictionary *)partsList forDevice:(MTRDevice *)device
+{
+    // Initialize the endpoint array with endpoint 0.
+    NSMutableArray<NSNumber *> * endpoints = [NSMutableArray arrayWithObject:@0];
+
+    [endpoints addObjectsFromArray:[device arrayOfNumbersFromAttributeValue:[device _dataValueWithoutDataVersion:partsList]]];
+    return endpoints;
+}
+
+- (void)test011_testDataStorageUpdatesWhenRemovingEndpoints
+{
+    NSNumber * deviceID = @(17);
+    __auto_type * device = [self getMTRDevice:deviceID];
+    __auto_type queue = dispatch_get_main_queue();
+    __auto_type * delegate = [[MTRDeviceTestDelegate alloc] init];
+    __auto_type * controller = device.deviceController;
+
+    XCTestExpectation * subscriptionExpectation = [self expectationWithDescription:@"Subscription has been set up"];
+
+    __block unsigned attributeReportsReceived = 0;
+    __block NSNumber * dataVersionForPartsList;
+    __block NSNumber * rootEndpoint = @0;
+
+    // This test will do the following -
+    // 1. Get the data version and attribute value of the parts list for endpoint 0 to inject a fake report. The attribute report will delete endpoint 2.
+    //    That should cause the endpoint and its corresponding clusters to be removed from data storage.
+    // 2. The data store is populated with cluster index and cluster data for endpoints 0, 1 and 2 initially.
+    // 3. After the fake attribute report is injected with deleted endpoint 2, make sure the data store is still populated with cluster index and cluster data
+    // for endpoints 0 and 1 but not 2.
+    __block NSDictionary * testDataForPartsList;
+    __block NSMutableArray * testClusterDataValueForPartsList;
+    delegate.onAttributeDataReceived = ^(NSArray<NSDictionary<NSString *, id> *> * attributeReport) {
+        attributeReportsReceived += attributeReport.count;
+        XCTAssert(attributeReportsReceived > 0);
```

Comment:
Why do we need a block var if we never reference it outside the block?

Also, please use XCTAssertGreaterThan.

Same in all the other places this code exists.

Reply:
done.